File: Python.PassDetection/src/features/velocities_dominant_foot.py
from typing import List

from src.domain.common import Vector3
from src.domain.inferences import FeatureCalculator, Feature
from src.domain.recordings import InputData


class VelocitiesDominantFootCalculator(FeatureCalculator):
    @property
    def size(self) -> int:
        return 3  # Each feature now has only one component (x, y, or z)

    def calculate(self, input_data: InputData) -> List[Feature]:
        dominant_positions = input_data.user_dominant_foot_positions
        timestamps = input_data.timestamps
        velocities = []

        for i in range(1, len(dominant_positions)):
            dt = timestamps[i] - timestamps[i - 1]
            if dt == 0:
                dt = 1e-6  # Avoid division by zero
            dx = (dominant_positions[i].x - dominant_positions[i - 1].x) / dt
            dy = (dominant_positions[i].y - dominant_positions[i - 1].y) / dt
            dz = (dominant_positions[i].z - dominant_positions[i - 1].z) / dt
            velocities.append(Vector3(dx, dy, dz))

        velocities.insert(0, Vector3(0, 0, 0))  # Insert a zero velocity at the start

        x_values = [vel.x for vel in velocities]
        y_values = [vel.y for vel in velocities]
        z_values = [vel.z for vel in velocities]

        return [
            Feature(name="velocities_dominant_foot_x", values=x_values),
            Feature(name="velocities_dominant_foot_y", values=y_values),
            Feature(name="velocities_dominant_foot_z", values=z_values),
        ]

File: Unity.Interactions/Assets/_Project/PassDetection.Replay/Features/VelocitiesDominantFoot.cs
using System.Collections.Generic;
using System.Linq;
using _Project.PassDetection.Common;
using _Project.PassDetection.Replay;
using _Project.PassDetection.Replay.Features;
using Tactive.MachineLearning._Project.MachineLearning;
using Tactive.MachineLearning.Features;
using UnityEngine;

namespace Src.Domain.Inferences
{
	public class VelocitiesDominantFoot : BaseFeature<InputData>
	{
		public override int Size => 3;

		public override List<Target> Calculate(InputData inputData)
		{
			var dominantPositions = inputData.UserDominantFootPositions;
			var timestamps = inputData.Timestamps;
			var velocities = new List<Vector3>();

			for (var i = 1; i < dominantPositions.Count; i++)
			{
				var dt = timestamps[i] - timestamps[i - 1];
				if (dt == 0) dt = 1e-6f;

				var dx = (dominantPositions[i].x - dominantPositions[i - 1].x) / dt;
				var dy = (dominantPositions[i].y - dominantPositions[i - 1].y) / dt;
				var dz = (dominantPositions[i].z - dominantPositions[i - 1].z) / dt;
				velocities.Add(new Vector3(dx, dy, dz));
			}

			velocities.Insert(0, Vector3.zero);

			var xValues = velocities.Select(v => v.x).ToList();
			var yValues = velocities.Select(v => v.y).ToList();
			var zValues = velocities.Select(v => v.z).ToList();

			return new List<Target>
			{
				new("velocities_dominant_foot_x", xValues),
				new("velocities_dominant_foot_y", yValues),
				new("velocities_dominant_foot_z", zValues)
			};
		}
	}
}